### ISSUE TITLE
Segfault when num RT args is too large

### ISSUE THREAD

**Issue by cglagovichTT** [tenstorrent/tt-metal#15263]
I've found a segfault during Llama development. In the model, I would get a segfault when the SDPA op hit the program cache for the first time. The segfault only occurred for specific input shapes and when the program cache is enabled. It occurs whether or not async is enabled.

## Early debugging
It seems that the segfault is caused by runtime arguments. I noticed that the shapes which led to a segfault caused the op to create more RT args than shapes which did not segfault. 
I proved to myself that the total number of RT args has something to do with the segfault by padding RT args in my passing test case. If I added enough extra RT args, the originally passing test would segfault.

I collected logs from valgrind for my passing and failing SDPA shapes. They are attached here. The failing log shows that in the sdpa program factory (when updating RT args, I believe) there are 31 invalid writes to memory. After these invalid writes, the heap metadata is corrupted. It is probably not a coincidence that there are exactly 31 RT args which are overridden in the SDPA op.  [valgrind_pass.log](https://github.com/user-attachments/files/17834075/valgrind_pass.log)
[valgrind_fail.log](https://github.com/user-attachments/files/17834076/valgrind_fail.log)

## Minimal repro
To repro this without the SDPA op, I have the following test on branch. I modified the binary program factory to create extra RT args and repro the segfault. This test will ND pass, segfault, or hang. If you run it a few times you should see the segfault.
```python
def test_fail_program_cache(device):
    device.enable_program_cache()

    """
    This test will run two different ops. The first op has been modified to increase the total number of rt args
    """

    a = torch.ones(32, 1, 32, 32)
    b = torch.ones(32, 1, 32, 32)

    a = ttnn.from_torch(a, device=device, layout=ttnn.TILE_LAYOUT, memory_config=ttnn.DRAM_MEMORY_CONFIG)
    b = ttnn.from_torch(b, device=device, layout=ttnn.TILE_LAYOUT, memory_config=ttnn.DRAM_MEMORY_CONFIG)

    for i in range(10):
        print(f"{i=}")
        c = ttnn.mul(a, b)
        d = ttnn.exp(c)
        ttnn.synchronize_device(device)

```

### Minimal repro
on branch [`cglagovich/segfaul`](https://github.com/tenstorrent/tt-metal/tree/cglagovich/segfaul)
```bash
pytest -svv tests/test_fail_program_cache.py
```

Expected output:
```
i=0
                 Always | INFO     | num_total_args: 26368
                 Always | INFO     | num_cores_total: 64
i=1
                 Always | INFO     | num_cores_total: 64
free(): invalid pointer
Fatal Python error: Aborted

Current thread 0x00007f1924253740 (most recent call first):
  File "/home/cglagovich/tt-metal/ttnn/ttnn/decorators.py", line 329 in __call__
  File "/home/cglagovich/tt-metal/tests/test_fail_program_cache.py", line 19 in test_fail_program_cache
```

### More complicated repro
This is the SDPA test I also isolated the segfault to.
```bash
WH_ARCH_YAML=wormhole_b0_80_arch_eth_dispatch.yaml pytest -svv tests/tt_eager/python_api_testing/unit_testing/misc/test_scaled_dot_product_attention_decode.py::test_fail
```

Expected output:
```
2024-11-20 16:55:22.407 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_scaled_dot_product_attention_decode:test_fail:220 - Testing with sequence length: 0
2024-11-20 16:55:22.407 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_scaled_dot_product_attention_decode:test_fail:221 - Using chunk size: 32
finished sdpa
finished exp
2024-11-20 16:55:23.726 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_scaled_dot_product_attention_decode:test_fail:220 - Testing with sequence length: 0
2024-11-20 16:55:23.726 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_scaled_dot_product_attention_decode:test_fail:221 - Using chunk size: 32
Fatal Python error: Segmentation fault
```

I'm assigning P0 because this issue blocks batch=32 Llama-Vision

**Comment by cglagovichTT** [comment]
fyi @uaydonat @caixunshiren @yieldthought

**Comment by cglagovichTT** [comment]
Potential hints:
- if you reduce `num_extra_args` which I added on branch, you can get rid of the segfault.
- I noticed that when there is a segfault, the memory addresses of the RT args vectors are different between the compile run and the program cache hit run

**Comment by tt-aho** [comment]
> Potential hints:
> 
> * if you reduce `num_extra_args` which I added on branch, you can get rid of the segfault.
> * I noticed that when there is a segfault, the memory addresses of the RT args vectors are different between the compile run and the program cache hit run

The memory addresses being different between compile/cache should be expected I think both in the seg fault case and not.

What happens with FD is that before dispatching for the first time, the memory addresses are to RTA vectors in the program. When we dispatch the first time, the RTA vectors gets copied to a new space and getting rtas now point into this space.

**Comment by cglagovichTT** [comment]
^ I got the addresses of the RT args from `GetRuntimeArgs`. Haven't those already been copied over to the "new space"?

**Comment by tt-aho** [comment]
Copy over to new space happens inside the first call to EnqueueProgram.
So ORTAs should always be the new space. I'm assuming getting the address in the compile pass this is done in the program factory code? If so then that's the old space

**Comment by cglagovichTT** [comment]
Thanks for explaining. Nevermind that potential hint, then

**Comment by tt-aho** [comment]
I pushed a fix to branch [aho/segfault](https://github.com/tenstorrent/tt-metal/commit/4a81d1df880b0de992a7fce676e3899f07aa1b81).

Issue here is that we incorrectly sized the vector we eventually copy the rtas to. We store the new addresses as we populate the vector, so since the vector was undersized it will get resized as we push into it, resulting in the original cached addresses becoming invalid.

**Comment by cglagovichTT** [comment]
Great! Do both of the repros pass?

**Comment by tt-aho** [comment]
Yes, I've also added a TT_ASSERT to check if we ever fail to correctly reserve the vector size as well,, so it should now have a clear message instead of segfaulting (in debug mode)
